homeassistant/components/roborock/__init__.py

```diff
@@ -29,66 +32,107 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     user_data = UserData.from_dict(entry.data[CONF_USER_DATA])
     api_client = RoborockApiClient(entry.data[CONF_USERNAME], entry.data[CONF_BASE_URL])
     _LOGGER.debug("Getting home data")
-    home_data = await api_client.get_home_data(user_data)
+    try:
+        home_data = await api_client.get_home_data(user_data)
+    except RoborockException as err:
+        raise ConfigEntryNotReady("Failed getting Roborock home_data.") from err
     _LOGGER.debug("Got home data %s", home_data)
     device_map: dict[str, HomeDataDevice] = {
         device.duid: device for device in home_data.devices + home_data.received_devices
     }
-    product_info = {product.id: product for product in home_data.products}
-    # Create a mqtt_client, which is needed to get the networking information of the device for local connection and in the future, get the map.
-    mqtt_clients = {
-        device.duid: RoborockMqttClient(
-            user_data, DeviceData(device, product_info[device.product_id].model)
-        )
-        for device in device_map.values()
-    }
-    network_results = await asyncio.gather(
-        *(mqtt_client.get_networking() for mqtt_client in mqtt_clients.values())
-    )
-    network_info = {
-        device.duid: result
-        for device, result in zip(device_map.values(), network_results)
-        if result is not None
+    product_info: dict[str, HomeDataProduct] = {
+        product.id: product for product in home_data.products
     }
-    if not network_info:
-        raise ConfigEntryNotReady(
-            "Could not get network information about your devices"
-        )
-    coordinator_map: dict[str, RoborockDataUpdateCoordinator] = {}
-    for device_id, device in device_map.items():
-        coordinator_map[device_id] = RoborockDataUpdateCoordinator(
-            hass,
-            device,
-            network_info[device_id],
-            product_info[device.product_id],
-            mqtt_clients[device.duid],
-        )
-    await asyncio.gather(
-        *(coordinator.verify_api() for coordinator in coordinator_map.values())
-    )
-    # If one device update fails - we still want to set up other devices
-    await asyncio.gather(
-        *(
-            coordinator.async_config_entry_first_refresh()
-            for coordinator in coordinator_map.values()
-        ),
-        return_exceptions=True,
+    # Get a Coordinator if the device is available or if we have connected to the device before
+    coordinators = await asyncio.gather(
+        *build_setup_functions(hass, device_map, user_data, product_info)
     )
+    # Valid coordinators are those where we had networking cached or we could get networking
+    valid_coordinators: list[RoborockDataUpdateCoordinator] = [
+        coord for coord in coordinators if coord is not None
+    ]
+    if len(valid_coordinators) == 0:
```

Comment:
Why is it ok to proceed when one of the coordinators fails? It seems to indicate the integration won't really be fully working. 

If i have two devices and one isn't setup then it seems like it will ignore it and never attempt to reconnect yet proceed and setup successfully. (I could see going with this approach if it still created entities and marked them unavailable, but i didn't necessarily see that is the case)

Reply:
I #97311 that's what I do - i always add the entity and I mark it as available - but it requires a bit of extra stuff including storing to determine what entities to add, etc. So I broke this up into a seperate PR so it was smaller and easy to get approved.

Basically the root of the issue is that some users have multiple devices on their account - and when one account is offline, the entire integration currently fails

Reply:
Is this a valid intermediate step though? I'm not sure that failing and dropping a device is better than failing.  I wondering if there is another intermediate step that preserves the behavior while still improving the error handling still with a partial refactor.

(I definitely see this myself, fwiw, so i understand the motivation)

Reply:
I'll put some thought into it tomorrow and see what I can do

Reply:
@allenporter Are you aware of anyway of just marking all entities on a device as unavailable without creating them? I can't create the entities when the device is offline without knowing which entities are supported by the device (I do this with a Storage obj in other PR) - but I don't think I can do it without it. If I didn't add the entities then the user would get 'this entity is no longer provided by this integration' which wouldn't be accurate. Perhaps I need to just close out this PR and just focus on getting the other PR added.

Reply:
I guess maybe i'm over thinking this. If they have unique ids maybe they'll show up as missing...

Reply:
I only have one device, but if it is offline, it does show up - just as missing, which I get is not ideal - but may be sufficient?

Reply:
<img width="755" alt="Screenshot 2023-11-19 at 10 00 51 PM" src="https://github.com/home-assistant/core/assets/20257911/b0d2fd17-824f-482f-b8f9-d62505788e55">

Reply:
Yes, agreed, that seems fine. Sorry for the false alarm there.

What do you think about passing `return_exceptions=True` to the `await asyncio.gather` and handling all the exceptions in the setup rather than swallowing them down inside the functions?

Reply:
Well the tricky part is I need to handle them per each device and I think it may add more confusion since I'd have to iterate through the results and if there was an exception, I need to also iterate through the original list so that I can get the information about the device for clear logs. I'm open to either way - but I thought I'd explain my logic.

Reply:
My impression is you either get back a coordinator or an exception, and then either (1) the exception could contain everything needed for logging or (2) the logging stays where it is and just raises.

This isn't a big deal, and my thought is just separation of responsibilities: the code for setting up a device doesn't have to know about the behavior here -- it just sets up a device and raises if needed. Then the setup code decides not to raise if certain circumstances (e.g. if any succeedes it proceeds anyway).

Reply:
Something like the change I just made? Is that what you were thinking?